src/main/java/org/apache/maven/plugins/gpg/FilesCollector.java

```diff
@@ -0,0 +1,178 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.maven.plugins.gpg;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.maven.artifact.Artifact;
+import org.apache.maven.plugin.MojoExecutionException;
+import org.apache.maven.plugin.MojoFailureException;
+import org.apache.maven.plugin.logging.Log;
+import org.apache.maven.project.MavenProject;
+import org.codehaus.plexus.util.FileUtils;
+import org.codehaus.plexus.util.SelectorUtils;
+
+/**
+ * Collects project artifact, the POM, and attached artifacts to be signed.
+ *
+ * @since 3.1.0
+ */
+public class FilesCollector {
+    private final MavenProject project;
+
+    private static final String DEFAULT_EXCLUDES[] =
+            new String[] {"**/*.md5", "**/*.sha1", "**/*.sha256", "**/*.sha512", "**/*.asc"};
+
+    private final String[] excludes;
+
+    private final Log log;
+
+    public FilesCollector(MavenProject project, String[] excludes, Log log) {
+        this.project = project;
+        this.log = log;
+        if (excludes == null || excludes.length == 0) {
+            this.excludes = DEFAULT_EXCLUDES;
+            return;
+        }
+        String newExcludes[] = new String[excludes.length];
+        for (int i = 0; i < excludes.length; i++) {
+            String pattern;
+            pattern = excludes[i].trim().replace('/', File.separatorChar).replace('\\', File.separatorChar);
+            if (pattern.endsWith(File.separator)) {
+                pattern += "**";
+            }
+            newExcludes[i] = pattern;
+        }
+        this.excludes = newExcludes;
+    }
+
+    public List<Item> collect() throws MojoExecutionException, MojoFailureException {
+        List<Item> items = new ArrayList<>();
+
+        if (!"pom".equals(project.getPackaging())) {
+            // ----------------------------------------------------------------------------
+            // Project artifact
+            // ----------------------------------------------------------------------------
+
+            Artifact artifact = project.getArtifact();
+
+            File file = artifact.getFile();
+
+            if (file != null && file.isFile()) {
+                items.add(new Item(file, artifact.getArtifactHandler().getExtension()));
+            } else if (project.getAttachedArtifacts().isEmpty()) {
+                throw new MojoFailureException("The project artifact has not been assembled yet. "
```

Comment:
This also applies to the main artifact I guess?

Reply:
1. AFAIK it is the main artifact: I don't get what you mean by "also applies"
2. this is purely an extract of existing algorithm from `gpg:sign`: the initial code just signs when collecting, when I differentiate the collection as a first step to let different signature algorithms after

see https://github.com/apache/maven-gpg-plugin/blob/525508093293cd1bb78d439cb28f2c48d7d79d9b/src/main/java/org/apache/maven/plugins/gpg/GpgSignAttachedMojo.java#L129

Reply:
Yes, correct. My bad.